service/prometheus/service.go

```diff
@@ -2,71 +2,230 @@ package prometheus
-	Logger micrologger.Logger
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
 
 	Address string
+	// ConfigMapKey is the key in the configmap under which the prometheus configuration is held.
+	ConfigMapKey       string
+	ConfigMapName      string
+	ConfigMapNamespace string
 }
 
 func DefaultConfig() Config {
 	return Config{
-		Logger: nil,
+		K8sClient: nil,
+		Logger:    nil,
 
-		Address: "",
+		Address:            "",
+		ConfigMapKey:       "",
+		ConfigMapName:      "",
+		ConfigMapNamespace: "",
 	}
 }
 
 type Service struct {
-	logger micrologger.Logger
-
-	// address is concatenated with the reload path in New.
-	address string
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+
+	// address is the address of the Prometheus instance we manage.
+	address            string
+	configMapKey       string
+	configMapName      string
+	configMapNamespace string
 }
 
 func New(config Config) (*Service, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.K8sClient must not be empty")
+	}
 	if config.Logger == nil {
 		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
 	}
 
 	if config.Address == "" {
 		return nil, microerror.Maskf(invalidConfigError, "config.Address must not be empty")
 	}
-
-	u, err := url.ParseRequestURI(config.Address)
-	if err != nil {
-		return nil, microerror.Maskf(invalidConfigError, "config.Address is invalid: %s", err)
+	if config.ConfigMapKey == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.ConfigMapKey must not be empty")
+	}
+	if config.ConfigMapName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.ConfigMapName must not be empty")
+	}
+	if config.ConfigMapNamespace == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.ConfigMapNamespace must not be empty")
 	}
-	u.Path = path.Join(u.Path, prometheusReloadPath)
 
 	service := &Service{
-		logger: config.Logger,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
 
-		address: u.String(),
+		address:            config.Address,
+		configMapKey:       config.ConfigMapKey,
+		configMapName:      config.ConfigMapName,
+		configMapNamespace: config.ConfigMapNamespace,
 	}
 
 	return service, nil
 }
 
 func (s *Service) Reload() error {
-	s.logger.Log("debug", fmt.Sprintf("reloading prometheus config: %s", s.address))
+	reloadRequired, err := s.isReloadRequired()
+	if err != nil {
+		return microerror.Maskf(reloadError, err.Error())
+	}
+
+	if reloadRequired {
+		if err := s.reload(); err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) isReloadRequired() (bool, error) {
+	s.logger.Log("debug", fmt.Sprintf("checking if reload is required"))
+
+	configurationReloadCheckCount.Inc()
+
+	currentConfigMapData, err := s.getCurrentConfigMapData()
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	currentConfiguration, err := s.getCurrentConfig()
```

Comment:
I would have called these methods `getConfigFromPrometheus` and `getConfigFromKubernetes` or something.  The current naming is not that straight forward from my current point of view.